src/main/java/org/omnifaces/util/JNDIObjectLocator.java

```diff
@@ -360,7 +223,221 @@ private InitialContext createInitialContext() {
+                private ArrayList<String> environmentKey;
+                private ArrayList<String> environmentValue;
+                private boolean namespaceSet;
+                private String namespaceValue;
+                private boolean noCaching;
+                private boolean cacheRemote;
+
+                JNDIObjectLocatorBuilder() {
+                }
+
+                /**
+		 * Adds an environment property.
+		 * @param environmentKey The key of the new environment property.
+		 * @param environmentValue The value of the new environment property.
+		 * @return This builder.
+		 * @throws NullPointerException When key or value is null.
+		 * @throws IllegalStateException When environment property is already set in this builder or when this builder is already build.
+		 */
+                public JNDIObjectLocatorBuilder environment(final String environmentKey, final String environmentValue) {
+                        requireNonNull(environmentKey, "key");
+                        requireNonNull(environmentValue, "value");
+                        if (this.environmentKey == null) {
+                                this.environmentKey = new ArrayList<>();
+                                this.environmentValue = new ArrayList<>();
+                        }
+                        this.environmentKey.add(environmentKey);
+                        this.environmentValue.add(environmentValue);
+                        return this;
+                }
+
+        	/**
+		 * Specifies the environment to be passed into {@link InitialContext}. The default is <code>null</code>.
+		 * @param environment The environment.
+		 * @return This builder.
+		 * @throws NullPointerException When given environment is null.
+		 * @throws IllegalStateException When environment is already set in this builder or when this builder is already build.
+		 */
+                public JNDIObjectLocatorBuilder environment(final Map<? extends String, ? extends String> environment) {
+                        if (environment == null) {
+                                throw new NullPointerException("environment cannot be null");
+                        }
+                        if (this.environmentKey == null) {
+                                this.environmentKey = new ArrayList<>();
+                                this.environmentValue = new ArrayList<>();
+                        }
+                        for (final Entry<? extends String, ? extends String> entry : environment.entrySet()) {
+                                this.environmentKey.add(entry.getKey());
+                                this.environmentValue.add(entry.getValue());
+                        }
+                        return this;
+                }
+
+                /**
+                 * Clears all environment
+                 * @return this
+                 */
+                public JNDIObjectLocatorBuilder clearEnvironment() {
```

Comment:
Doesn't look like clean API. What's the use case of this?

Reply:
This if from the 'builder pattern' literature, I can remove it if you want, I don't use it

Reply:
Okay. But all these changes make the builder less robust. I do personally not agree with that. You should not assume that all endusers are smart developers. The API should be able to error out with a runtime exception when endusers are using the API in a wrong way. Keep in mind that this all ends up as a public API not as an internal API. 

Why didn't you just edit the existing constructors?

Reply:
I didn't write this. This is delombok'd code. It says so in the comments. Builder pattern doesn't have to be 'perfect' it should just be flexible and work. This is why it's on the bottom and not the top.

I can remove this if that bothers you so much, that's not a problem

Reply:
I think I see what you are getting at, however I disagree, having had plenty of experience with builder pattern.
Builder is a special case. The premise is that builders could be passed around unrelated code,
and some methods could 'reset' what other methods are doing.
This scenario would need to be able to set / reset same variables multiple times.
This is why `clearEnvirontment()` exists and it's why methods don't throw exceptions if they were already set.
I did not invent any of this, just following what Lombok generates / does, and they follow what the literature says.